Sources/NIOPosix/Bootstrap.swift

```diff
@@ -1352,6 +1352,483 @@ public final class ClientBootstrap: NIOClientTCPBootstrapProtocol {
+                        outboundType: channelOutboundType
+                    )
+                }
+            },
+            postRegisterTransformation: { result in
+                eventLoop.makeSucceededFuture(result)
+            },
+            body
+        )
+    }
+
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    private func initializeAndRegisterChannel<ChannelInbound: Sendable, ChannelOutbound: Sendable>(
+        channel: SocketChannel,
+        channelInboundType: ChannelInbound.Type = ChannelInbound.self,
+        channelOutboundType: ChannelOutbound.Type = ChannelOutbound.self,
+        channelBackpressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark? = nil,
+        isOutboundHalfClosureEnabled: Bool = false,
+        _ body: @escaping (Channel) -> EventLoopFuture<Void>
+    ) -> EventLoopFuture<NIOAsyncChannel<ChannelInbound, ChannelOutbound>> {
+        return self.initializeAndRegisterChannel(
+            channel: channel
+        ) { channel in
+            channel.eventLoop.makeCompletedFuture {
+                return try NIOAsyncChannel(
+                    synchronouslyWrapping: channel,
+                    backpressureStrategy: channelBackpressureStrategy,
+                    isOutboundHalfClosureEnabled: isOutboundHalfClosureEnabled,
+                    inboundType: channelInboundType,
+                    outboundType: channelOutboundType
+                )
+            }
+        } registration: { channel in
+            channel.registerAndDoSynchronously(body)
+        } postRegisterTransformation: { result in
+            channel.eventLoop.makeSucceededFuture(result)
+        }
+    }
+}
+
+// MARK: AsyncChannel based connect with protocol negotiation
+
+extension ClientBootstrap {
+    /// Specify the `host` and `port` to connect to for the TCP `Channel` that will be established.
+    ///
+    /// - Parameters:
+    ///   - host: The host to connect to.
+    ///   - port: The port to connect to.
+    ///   - channelInitializer: A closure to initialize the channel which must return the handler that is used for negotiating
+    ///   the protocol.
+    /// - Returns: The protocol negotiation result.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func connect<Handler: NIOProtocolNegotiationHandler>(
+        host: String,
+        port: Int,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Handler>
+    ) async throws -> Handler.NegotiationResult {
+        let eventLoop = self.group.next()
+        return try await self.connect(
+            host: host,
+            port: port,
+            eventLoop: eventLoop,
+            channelInitializer: channelInitializer,
+            postRegisterTransformation: { handler in
+                handler.protocolNegotiationResult.flatMap { result in
+                    ServerBootstrap.waitForFinalResult(result, eventLoop: eventLoop)
```

Comment:
Feels weird calling a `ServerBootstrap` function here; might make more sense to move the function to `NIOProtocolNegotiationResult` instead? 

```swift
extension NIOProtocolNegotiationResult {
 func resolve(on eventLoop: EventLoop) -> EventLoopFuture<NegotiationResult> { 
    // ...
  }
}
```